File: leetcode/algorithms/27_remove_element/main.py
from typing import List


class RemoveElement:
    """
    # Two Pointers
    #
    # Complexities:
    #   N - Length of `nums`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(1)
    """
    def removeElement(self, nums: List[int], val: int) -> int:
        nums[:] = sorted(filter(lambda num: num != val, nums))

        return len(nums)


    # Solution
    """
    # Two Pointers
    #
    # Complexities:
    #   N - Length of `nums`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(1)
    """
    def solution(self, nums: List[int], val: int) -> int:
        k = 0

        for i in range(len(nums)):
            if nums[i] != val:
                nums[k] = nums[i]
                k += 1

        return k

File: leetcode/algorithms/27_remove_element/main.dart
class RemoveElement {
  /**
   * Two Pointers
   *
   * Complexities:
   *   N - Length of `nums`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(1)
   */
  int removeElement(List<int> nums, int val) {
    nums.removeWhere((e) => e == val);

    return 0;
  }


  // Solution
  /**
   * Two Pointers
   *
   * Complexities:
   *   N - Length of `nums`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(1)
   */
  int solution(List<int> nums, int val) {
    int k = 0;

    for (int i = 0; i < nums.length; i++) {
      if (nums[i] != val) {
        nums[k] = nums[i];
        k++;
      }
    }

    return k;
  }
}
